gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/ExternalDataConfiguration.java

```diff
@@ -0,0 +1,419 @@
+
+    /**
+     * Sets the fully-qualified URIs that point to your data in Google Cloud Storage. Each URI can
+     * contain one '*' wildcard character that must come after the bucket's name. Size limits
+     * related to load jobs apply to external data sources, plus an additional limit of 10 GB
+     * maximum size across all URIs.
+     *
+     * @see <a href="https://cloud.google.com/bigquery/loading-data-into-bigquery#quota">Quota</a>
+     */
+    public Builder sourceUris(List<String> sourceUris) {
+      this.sourceUris = ImmutableList.copyOf(checkNotNull(sourceUris));
+      return this;
+    }
+
+    /**
+     * Sets the schema for the external data.
+     */
+    public Builder schema(TableSchema schema) {
+      this.schema = checkNotNull(schema);
+      return this;
+    }
+
+    /**
+     * Sets the source format of the external data.
+     *
+     * <a href="https://cloud.google.com/bigquery/docs/reference/v2/tables#externalDataConfiguration.sourceFormat">
+     *     Source Format</a>
+     */
+    public Builder sourceFormat(String sourceFormat) {
+      this.sourceFormat = checkNotNull(sourceFormat);
+      return this;
+    }
+
+    /**
+     * Sets the maximum number of bad records that BigQuery can ignore when reading data. If the
+     * number of bad records exceeds this value, an invalid error is returned in the job result.
+     * The default value is 0, which requires that all records are valid.
+     */
+    public Builder maxBadRecords(Integer maxBadRecords) {
+      this.maxBadRecords = maxBadRecords;
+      return this;
+    }
+
+    /**
+     * Sets whether BigQuery should allow extra values that are not represented in the table schema.
+     * If true, the extra values are ignored. If false, records with extra columns are treated as
+     * bad records, and if there are too many bad records, an invalid error is returned in the job
+     * result. The default value is false. The value set with
+     * {@link #sourceFormat(String)} property determines what
+     * BigQuery treats as an extra value.
+     *
+     * @see <a href="https://cloud.google.com/bigquery/docs/reference/v2/tables#externalDataConfiguration.ignoreUnknownValues">
+     *     Ignore Unknown Values</a>
+     */
+    public Builder ignoreUnknownValues(Boolean ignoreUnknownValues) {
+      this.ignoreUnknownValues = ignoreUnknownValues;
+      return this;
+    }
+
+    /**
+     * Sets compression type of the data source. By default no compression is assumed.
+     *
+     * @see <a href="https://cloud.google.com/bigquery/docs/reference/v2/tables#externalDataConfiguration.compression">
+     *     Compression</a>
+     */
+    public Builder compression(String compression) {
+      this.compression = compression;
+      return this;
+    }
+
+    /**
+     * Sets additional properties to be used to parse CSV data (used when
+     * {@link #sourceFormat(String)} is set to CSV).
+     */
+    public Builder csvOptions(CsvOptions csvOptions) {
```

Comment:
The approach I adopted so far for *pick-one* fields is to have all fields in the top level class and design the builder and factory methods so that only one is initialized at a time. It would be something like:
```java
public static ExternalDataConfiguration of(String sourceUri, TableSchema schema, CsvOptions options) {
  return builder(sourceUri, schema, options).format("CSV").build();
}

public static ExternalDataConfiguration of(String sourceUri, TableSchema schema, JsonOptions options) {
  return builder(sourceUri, schema, options).format("NEWLINE_DELEMITED_JSON").build();
}
```

This should enforce a "good" construction.